### DISCUSSION TITLE
MapsForge with Yandex map (how use wgs84Mercator)

### DISCUSSION THREAD

**Discussion by AlVZhykov** [mapsforge/mapsforge#1284] [Ideas]
Good day,

mapsforge has a sphericalMercator. It works with OSM, Google, BING maps.

But there are companies providing other Mercator - wgs84Mercator. Here is an example (Yandex map): https://jsfiddle.net/utvndLjr/16/

When I download maps with another Mercator (wgs84Mercator) object position is displayed incorrectly.

example: correct position
centerLat=55.0307
centerLon=73.2695
![correct](https://user-images.githubusercontent.com/98458649/151148813-f275151b-3466-4436-85c3-314e52c646cf.png)

example: NOT correct position
centerLat=55.0307
centerLon=73.2695
![not correct](https://user-images.githubusercontent.com/98458649/151148819-d4766cac-94ee-47b6-9b75-cd0dff0d2bef.png)

This happens because it is used sphericalMercator in OSM maps. Yandex maps are used wgs84Mercator.

Please tell me how to use the wgs84Mercator in mapsforge?

**Comment by devemux86** [comment]
Web Mercator is the standard in mapping applications:
https://en.wikipedia.org/wiki/Web_Mercator_projection

**Reply by AlVZhykov** [reply]
Thank you. 

I need use Spherical Mercator EPSG:3857 for OSM or Google maps and wgs84Mercator (ellipsoidal) for Yandex map.

I found only such a solution:
public class Tile {

    public static void main(String[] args) {
        int zoom = 10;
        double lat = 55.0307d;
        double lon = 73.2695d;

        getTileOSM(lat, lon, zoom);
        getTileYandex(lat, lon, zoom);
    }

    public static void getTileOSM(final double lat, final double lon, final int zoom) {
        int xtile = (int)Math.floor( (lon + 180) / 360 * (1<<zoom));
        System.out.println("x: " + xtile);
        int ytile = (int)Math.floor( (1 - Math.log(Math.tan(Math.toRadians(lat)) + 1 / Math.cos(Math.toRadians(lat))) / Math.PI) / 2 * (1<<zoom));
        System.out.println("y: " + ytile);
        if (xtile < 0)
            xtile=0;
        if (xtile >= (1<<zoom))
            xtile=((1<<zoom)-1);
        if (ytile < 0)
            ytile=0;
        if (ytile >= (1<<zoom))
            ytile=((1<<zoom)-1);
        System.out.println("https://tile.openstreetmap.org/" + "" + zoom + "/" + xtile + "/" + ytile + ".png");
    }

    // Яндекс алгоритм взят с сайта https://jsfiddle.net/utvndLjr/16/
    public static void getTileYandex(final double lat, final double lon, final int z) {
        double x_p, y_p;
        double rho,
                pi = Math.PI,
                beta,
                phi,
                theta,
                e = 0.0818191908426; // eccentricity эксцентриситет

        rho = Math.pow(2, z + 8) / 2;
        beta = lat * pi / 180;
        phi = (1 - e * Math.sin(beta)) / (1 + e * Math.sin(beta));
        theta = Math.tan(pi / 4 + beta / 2) * Math.pow(phi, e / 2);

        x_p = rho * (1 + lon / 180);
        y_p = rho * (1 - Math.log(theta) / pi);

        System.out.println("Yandex");
        System.out.println("x: " + x_p / 256);
        System.out.println("y: " + y_p / 256);
        System.out.println("https://core-renderer-tiles.maps.yandex.net/tiles?l=map&x="
                + (int)(x_p / 256) + "&y=" + (int)(y_p / 256) + "&z=" + z + "&scale=1&lang=ru_RU");
    }
}

So when I have latlong, I can calculate x, y for Spherical Mercator, or Ellipsoidal Mercator.

**Reply by demiantres** [reply]
Are you getting the correct Yandex tiles with this? Or are they still (at least a bit) off?

**Reply by AlVZhykov** [reply]
I had 2 way to solve the problem of incorrect display of objects:

1. Download other tiles from Yandex. But this path is unsuccessful.

I do not know how to change MercatorProjection class (package org.mapsforge.core.util;)

I can't change mapsforge libraries. I wanted to  download  (with  lib mapsforge) other tiles for Yandex map.

2. And I'm trying to implement another way. Change the method DRAW  in my class (extended Marker) for objects:
getTileSourceComboBox - this is the type of card

    @Override
    public synchronized void draw(BoundingBox boundingBox, byte zoomLevel, Canvas canvas, Point topLeftPoint) {

......

        if (MapPane.getTileSourceComboBox().equals("YandexMapOnline")) {
            // Wgs84Mercator
            int[] intMasWgs84 = MercatorTools.getWgs84MercatorValue(getLatLong(), zoomLevel);
            pixelX = intMasWgs84[0];
            pixelY = intMasWgs84[1];
        } else {
            // Default
            long mapSize = MercatorProjection.getMapSize(zoomLevel, this.displayModel.getTileSize());
            pixelX = MercatorProjection.longitudeToPixelX(getLatLong().longitude, mapSize);
            pixelY = MercatorProjection.latitudeToPixelY(getLatLong().latitude, mapSize);
        }
.......

}
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

public class MercatorTools {

    public static int[] getWgs84MercatorValue(LatLong latLong, int zoom) {

        int x, y;
        double  lat, lon,
                rho,
                pi = Math.PI,
                beta,
                phi,
                theta,
                e = 0.0818191908426; // эксцентриситет

        lon = latLong.longitude;
        lat = latLong.latitude;

        rho = Math.pow(2, zoom + 8) / 2;
        beta = lat * pi / 180;
        phi = (1 - e * Math.sin(beta)) / (1 + e * Math.sin(beta));
        theta = Math.tan(pi / 4 + beta / 2) * Math.pow(phi, e / 2);

        x = (int)(rho * (1 + lon / 180));
        y = (int)(rho * (1 - Math.log(theta) / pi));

        return new int[]{x, y};
    }
}


I was able to implement the correct display of markers:
![OSM](https://user-images.githubusercontent.com/98458649/151308114-e579cbb1-d3d5-483c-823c-f9550d050fd6.png)
![Yandex](https://user-images.githubusercontent.com/98458649/151308116-2d5fcf98-5d61-4c14-90ee-b309caf9119d.png)

But it wasn't enough:
![1](https://user-images.githubusercontent.com/98458649/151308419-2043f490-8313-442d-8c29-4b3c81a15c91.png)
![2](https://user-images.githubusercontent.com/98458649/151308424-9670936f-8d85-4dd4-96d8-6bcaa7b7d946.png)

Now you need to change the draw method in Rectangle, Circle and Polyline :(

I haven't come up with a good solution yet

**Reply by demiantres** [reply]
From my point of view you have two options:

1. Add a new projection to the Mapsforge project (as you are trying to do?). This seems to be a lot of work and prone to errors.
2. Transform the tiles to WGS84 projection.

I have done 2.) with some other (non-Mapsforge) project but in Mapsforge its the same:

1. Get the WGS84 tile's bounding box and convert it to the other projection (e.g. using proj4).
2. Convert the projected bounding box to tile coordinates. Usually you will get multiple tiles due to overlaps!
3. Download the tiles.
4. Project the tiles back to WGS84 and create a WGS84 tile by stitching the other tiles.

**Reply by AlVZhykov** [reply]
Thanks,

I wanted to add another class (analogous to Mercator Projection), but Mercator Projection is used in many other classes:
MapPositionUtil, LayerUtil, MapViewProjection and other classes. It turned out to be very difficult.

I decided to change the projection in yandex maps through transformation:
1. In the SAS.Planet program (sasgis.org) I downloaded a section of the map in GeoTIFF format
2. In another QGIS program (qgis.org ) I changed the map projection from EPSG 3395 to EPSG 3785
3. In the QGIS program, using the QTiles module, I cut the modified map into tiles.

As a result, it turned out not very good image quality, the names are blurry :(But unfortunately I couldn't think of a better option to use MapsForge with Yandex maps.
![1](https://user-images.githubusercontent.com/98458649/151914261-9f929cd2-f2fe-437c-8116-d2833fea291f.png)
